// @coti-io/coti-contracts/typechain-types/contracts/onboard/AccountOnboard.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface AccountOnboardInterface extends Interface {
  getFunction(nameOrSignature: "onboardAccount"): FunctionFragment;

  getEvent(nameOrSignatureOrTopic: "AccountOnboarded"): EventFragment;

  encodeFunctionData(
    functionFragment: "onboardAccount",
    values: [BytesLike, BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "onboardAccount",
    data: BytesLike
  ): Result;
}

export namespace AccountOnboardedEvent {
  export type InputTuple = [_from: AddressLike, userKey: BytesLike];
  export type OutputTuple = [_from: string, userKey: string];
  export interface OutputObject {
    _from: string;
    userKey: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface AccountOnboard extends BaseContract {
  connect(runner?: ContractRunner | null): AccountOnboard;
  waitForDeployment(): Promise<this>;

  interface: AccountOnboardInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  onboardAccount: TypedContractMethod<
    [publicKey: BytesLike, signedEK: BytesLike],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "onboardAccount"
  ): TypedContractMethod<
    [publicKey: BytesLike, signedEK: BytesLike],
    [void],
    "nonpayable"
  >;

  getEvent(
    key: "AccountOnboarded"
  ): TypedContractEvent<
    AccountOnboardedEvent.InputTuple,
    AccountOnboardedEvent.OutputTuple,
    AccountOnboardedEvent.OutputObject
  >;

  filters: {
    "AccountOnboarded(address,bytes)": TypedContractEvent<
      AccountOnboardedEvent.InputTuple,
      AccountOnboardedEvent.OutputTuple,
      AccountOnboardedEvent.OutputObject
    >;
    AccountOnboarded: TypedContractEvent<
      AccountOnboardedEvent.InputTuple,
      AccountOnboardedEvent.OutputTuple,
      AccountOnboardedEvent.OutputObject
    >;
  };
}
