// @coti-io/coti-contracts/typechain-types/contracts/access/DataPrivacyFramework/DataPrivacyFramework.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../../common";

export declare namespace DataPrivacyFramework {
  export type ConditionStruct = {
    id: BigNumberish;
    caller: AddressLike;
    operation: string;
    active: boolean;
    falseKey: boolean;
    trueKey: boolean;
    timestampBefore: BigNumberish;
    timestampAfter: BigNumberish;
    uintParameter: BigNumberish;
    addressParameter: AddressLike;
    stringParameter: string;
  };

  export type ConditionStructOutput = [
    id: bigint,
    caller: string,
    operation: string,
    active: boolean,
    falseKey: boolean,
    trueKey: boolean,
    timestampBefore: bigint,
    timestampAfter: bigint,
    uintParameter: bigint,
    addressParameter: string,
    stringParameter: string
  ] & {
    id: bigint;
    caller: string;
    operation: string;
    active: boolean;
    falseKey: boolean;
    trueKey: boolean;
    timestampBefore: bigint;
    timestampAfter: bigint;
    uintParameter: bigint;
    addressParameter: string;
    stringParameter: string;
  };

  export type InputDataStruct = {
    caller: AddressLike;
    operation: string;
    active: boolean;
    timestampBefore: BigNumberish;
    timestampAfter: BigNumberish;
    falseKey: boolean;
    trueKey: boolean;
    uintParameter: BigNumberish;
    addressParameter: AddressLike;
    stringParameter: string;
  };

  export type InputDataStructOutput = [
    caller: string,
    operation: string,
    active: boolean,
    timestampBefore: bigint,
    timestampAfter: bigint,
    falseKey: boolean,
    trueKey: boolean,
    uintParameter: bigint,
    addressParameter: string,
    stringParameter: string
  ] & {
    caller: string;
    operation: string;
    active: boolean;
    timestampBefore: bigint;
    timestampAfter: bigint;
    falseKey: boolean;
    trueKey: boolean;
    uintParameter: bigint;
    addressParameter: string;
    stringParameter: string;
  };
}

export interface DataPrivacyFrameworkInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "ADDRESS_ALL"
      | "STRING_ALL"
      | "activePermissions"
      | "addAllowedOperation"
      | "addRestrictedOperation"
      | "addressDefaultPermission"
      | "allowedOperations"
      | "conditions"
      | "getConditions"
      | "getConditionsCount"
      | "isOperationAllowed(address,string,uint256)"
      | "isOperationAllowed(address,string)"
      | "isOperationAllowed(address,string,string)"
      | "isOperationAllowed(address,string,address)"
      | "operationDefaultPermission"
      | "permissions"
      | "removeAllowedOperation"
      | "removeRestrictedOperation"
      | "restrictedOperations"
      | "setAddressDefaultPermission"
      | "setOperationDefaultPermission"
      | "setPermission"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "ADDRESS_ALL",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "STRING_ALL",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "activePermissions",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "addAllowedOperation",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "addRestrictedOperation",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "addressDefaultPermission",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "allowedOperations",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "conditions",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getConditions",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getConditionsCount",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "isOperationAllowed(address,string,uint256)",
    values: [AddressLike, string, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "isOperationAllowed(address,string)",
    values: [AddressLike, string]
  ): string;
  encodeFunctionData(
    functionFragment: "isOperationAllowed(address,string,string)",
    values: [AddressLike, string, string]
  ): string;
  encodeFunctionData(
    functionFragment: "isOperationAllowed(address,string,address)",
    values: [AddressLike, string, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "operationDefaultPermission",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "permissions",
    values: [AddressLike, string]
  ): string;
  encodeFunctionData(
    functionFragment: "removeAllowedOperation",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "removeRestrictedOperation",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "restrictedOperations",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "setAddressDefaultPermission",
    values: [boolean]
  ): string;
  encodeFunctionData(
    functionFragment: "setOperationDefaultPermission",
    values: [boolean]
  ): string;
  encodeFunctionData(
    functionFragment: "setPermission",
    values: [DataPrivacyFramework.InputDataStruct]
  ): string;

  decodeFunctionResult(
    functionFragment: "ADDRESS_ALL",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "STRING_ALL", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "activePermissions",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "addAllowedOperation",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "addRestrictedOperation",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "addressDefaultPermission",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "allowedOperations",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "conditions", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getConditions",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getConditionsCount",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isOperationAllowed(address,string,uint256)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isOperationAllowed(address,string)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isOperationAllowed(address,string,string)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isOperationAllowed(address,string,address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "operationDefaultPermission",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "permissions",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "removeAllowedOperation",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "removeRestrictedOperation",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "restrictedOperations",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setAddressDefaultPermission",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setOperationDefaultPermission",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setPermission",
    data: BytesLike
  ): Result;
}

export interface DataPrivacyFramework extends BaseContract {
  connect(runner?: ContractRunner | null): DataPrivacyFramework;
  waitForDeployment(): Promise<this>;

  interface: DataPrivacyFrameworkInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  ADDRESS_ALL: TypedContractMethod<[], [string], "view">;

  STRING_ALL: TypedContractMethod<[], [string], "view">;

  activePermissions: TypedContractMethod<[arg0: AddressLike], [bigint], "view">;

  addAllowedOperation: TypedContractMethod<
    [operation: string],
    [boolean],
    "nonpayable"
  >;

  addRestrictedOperation: TypedContractMethod<
    [operation: string],
    [boolean],
    "nonpayable"
  >;

  addressDefaultPermission: TypedContractMethod<[], [boolean], "view">;

  allowedOperations: TypedContractMethod<[arg0: string], [boolean], "view">;

  conditions: TypedContractMethod<
    [arg0: BigNumberish],
    [
      [
        bigint,
        string,
        string,
        boolean,
        boolean,
        boolean,
        bigint,
        bigint,
        bigint,
        string,
        string
      ] & {
        id: bigint;
        caller: string;
        operation: string;
        active: boolean;
        falseKey: boolean;
        trueKey: boolean;
        timestampBefore: bigint;
        timestampAfter: bigint;
        uintParameter: bigint;
        addressParameter: string;
        stringParameter: string;
      }
    ],
    "view"
  >;

  getConditions: TypedContractMethod<
    [startIdx: BigNumberish, chunkSize: BigNumberish],
    [DataPrivacyFramework.ConditionStructOutput[]],
    "view"
  >;

  getConditionsCount: TypedContractMethod<[], [bigint], "view">;

  "isOperationAllowed(address,string,uint256)": TypedContractMethod<
    [caller: AddressLike, operation: string, uintParameter: BigNumberish],
    [boolean],
    "view"
  >;

  "isOperationAllowed(address,string)": TypedContractMethod<
    [caller: AddressLike, operation: string],
    [boolean],
    "view"
  >;

  "isOperationAllowed(address,string,string)": TypedContractMethod<
    [caller: AddressLike, operation: string, stringParameter: string],
    [boolean],
    "view"
  >;

  "isOperationAllowed(address,string,address)": TypedContractMethod<
    [caller: AddressLike, operation: string, addressParameter: AddressLike],
    [boolean],
    "view"
  >;

  operationDefaultPermission: TypedContractMethod<[], [boolean], "view">;

  permissions: TypedContractMethod<
    [arg0: AddressLike, arg1: string],
    [bigint],
    "view"
  >;

  removeAllowedOperation: TypedContractMethod<
    [operation: string],
    [boolean],
    "nonpayable"
  >;

  removeRestrictedOperation: TypedContractMethod<
    [operation: string],
    [boolean],
    "nonpayable"
  >;

  restrictedOperations: TypedContractMethod<[arg0: string], [boolean], "view">;

  setAddressDefaultPermission: TypedContractMethod<
    [defaultPermission: boolean],
    [boolean],
    "nonpayable"
  >;

  setOperationDefaultPermission: TypedContractMethod<
    [defaultPermission: boolean],
    [boolean],
    "nonpayable"
  >;

  setPermission: TypedContractMethod<
    [inputData: DataPrivacyFramework.InputDataStruct],
    [boolean],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "ADDRESS_ALL"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "STRING_ALL"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "activePermissions"
  ): TypedContractMethod<[arg0: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "addAllowedOperation"
  ): TypedContractMethod<[operation: string], [boolean], "nonpayable">;
  getFunction(
    nameOrSignature: "addRestrictedOperation"
  ): TypedContractMethod<[operation: string], [boolean], "nonpayable">;
  getFunction(
    nameOrSignature: "addressDefaultPermission"
  ): TypedContractMethod<[], [boolean], "view">;
  getFunction(
    nameOrSignature: "allowedOperations"
  ): TypedContractMethod<[arg0: string], [boolean], "view">;
  getFunction(
    nameOrSignature: "conditions"
  ): TypedContractMethod<
    [arg0: BigNumberish],
    [
      [
        bigint,
        string,
        string,
        boolean,
        boolean,
        boolean,
        bigint,
        bigint,
        bigint,
        string,
        string
      ] & {
        id: bigint;
        caller: string;
        operation: string;
        active: boolean;
        falseKey: boolean;
        trueKey: boolean;
        timestampBefore: bigint;
        timestampAfter: bigint;
        uintParameter: bigint;
        addressParameter: string;
        stringParameter: string;
      }
    ],
    "view"
  >;
  getFunction(
    nameOrSignature: "getConditions"
  ): TypedContractMethod<
    [startIdx: BigNumberish, chunkSize: BigNumberish],
    [DataPrivacyFramework.ConditionStructOutput[]],
    "view"
  >;
  getFunction(
    nameOrSignature: "getConditionsCount"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "isOperationAllowed(address,string,uint256)"
  ): TypedContractMethod<
    [caller: AddressLike, operation: string, uintParameter: BigNumberish],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "isOperationAllowed(address,string)"
  ): TypedContractMethod<
    [caller: AddressLike, operation: string],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "isOperationAllowed(address,string,string)"
  ): TypedContractMethod<
    [caller: AddressLike, operation: string, stringParameter: string],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "isOperationAllowed(address,string,address)"
  ): TypedContractMethod<
    [caller: AddressLike, operation: string, addressParameter: AddressLike],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "operationDefaultPermission"
  ): TypedContractMethod<[], [boolean], "view">;
  getFunction(
    nameOrSignature: "permissions"
  ): TypedContractMethod<[arg0: AddressLike, arg1: string], [bigint], "view">;
  getFunction(
    nameOrSignature: "removeAllowedOperation"
  ): TypedContractMethod<[operation: string], [boolean], "nonpayable">;
  getFunction(
    nameOrSignature: "removeRestrictedOperation"
  ): TypedContractMethod<[operation: string], [boolean], "nonpayable">;
  getFunction(
    nameOrSignature: "restrictedOperations"
  ): TypedContractMethod<[arg0: string], [boolean], "view">;
  getFunction(
    nameOrSignature: "setAddressDefaultPermission"
  ): TypedContractMethod<[defaultPermission: boolean], [boolean], "nonpayable">;
  getFunction(
    nameOrSignature: "setOperationDefaultPermission"
  ): TypedContractMethod<[defaultPermission: boolean], [boolean], "nonpayable">;
  getFunction(
    nameOrSignature: "setPermission"
  ): TypedContractMethod<
    [inputData: DataPrivacyFramework.InputDataStruct],
    [boolean],
    "nonpayable"
  >;

  filters: {};
}
