// @coti-io/coti-ethers/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserProvider = exports.JsonRpcSigner = exports.JsonRpcApiProvider = exports.Wallet = exports.isGasEstimationValid = exports.validateGasEstimation = exports.recoverAesFromTx = exports.onboard = exports.getAccountOnboardContract = exports.isProviderConnected = exports.getLatestBlock = exports.printNetworkDetails = exports.getDefaultProvider = exports.transferNative = exports.getEoa = exports.getNativeBalance = exports.addressValid = exports.getNonce = exports.validateAddress = exports.getAccountBalance = exports.printAccountDetails = void 0;
__exportStar(require("ethers"), exports);
__exportStar(require("./utils/constants"), exports);
var utils_1 = require("./utils");
Object.defineProperty(exports, "printAccountDetails", { enumerable: true, get: function () { return utils_1.printAccountDetails; } });
Object.defineProperty(exports, "getAccountBalance", { enumerable: true, get: function () { return utils_1.getAccountBalance; } });
Object.defineProperty(exports, "validateAddress", { enumerable: true, get: function () { return utils_1.validateAddress; } });
Object.defineProperty(exports, "getNonce", { enumerable: true, get: function () { return utils_1.getNonce; } });
Object.defineProperty(exports, "addressValid", { enumerable: true, get: function () { return utils_1.addressValid; } });
Object.defineProperty(exports, "getNativeBalance", { enumerable: true, get: function () { return utils_1.getNativeBalance; } });
Object.defineProperty(exports, "getEoa", { enumerable: true, get: function () { return utils_1.getEoa; } });
Object.defineProperty(exports, "transferNative", { enumerable: true, get: function () { return utils_1.transferNative; } });
Object.defineProperty(exports, "getDefaultProvider", { enumerable: true, get: function () { return utils_1.getDefaultProvider; } });
Object.defineProperty(exports, "printNetworkDetails", { enumerable: true, get: function () { return utils_1.printNetworkDetails; } });
Object.defineProperty(exports, "getLatestBlock", { enumerable: true, get: function () { return utils_1.getLatestBlock; } });
Object.defineProperty(exports, "isProviderConnected", { enumerable: true, get: function () { return utils_1.isProviderConnected; } });
Object.defineProperty(exports, "getAccountOnboardContract", { enumerable: true, get: function () { return utils_1.getAccountOnboardContract; } });
Object.defineProperty(exports, "onboard", { enumerable: true, get: function () { return utils_1.onboard; } });
Object.defineProperty(exports, "recoverAesFromTx", { enumerable: true, get: function () { return utils_1.recoverAesFromTx; } });
Object.defineProperty(exports, "validateGasEstimation", { enumerable: true, get: function () { return utils_1.validateGasEstimation; } });
Object.defineProperty(exports, "isGasEstimationValid", { enumerable: true, get: function () { return utils_1.isGasEstimationValid; } });
var Wallet_1 = require("./wallet/Wallet");
Object.defineProperty(exports, "Wallet", { enumerable: true, get: function () { return Wallet_1.Wallet; } });
var JsonRpcApiProvider_1 = require("./providers/JsonRpcApiProvider");
Object.defineProperty(exports, "JsonRpcApiProvider", { enumerable: true, get: function () { return JsonRpcApiProvider_1.JsonRpcApiProvider; } });
var JsonRpcSigner_1 = require("./providers/JsonRpcSigner");
Object.defineProperty(exports, "JsonRpcSigner", { enumerable: true, get: function () { return JsonRpcSigner_1.JsonRpcSigner; } });
var BrowserProvider_1 = require("./providers/BrowserProvider");
Object.defineProperty(exports, "BrowserProvider", { enumerable: true, get: function () { return BrowserProvider_1.BrowserProvider; } });
__exportStar(require("./types"), exports);
