// @coti-io/coti-contracts/typechain-types/contracts/mocks/utils/mpc/index.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type * as comparison1TestsContractSol from "./Comparison1Tests.Contract.sol";
export type { comparison1TestsContractSol };
import type * as offboardToUserKeyTestsContractSol from "./OffboardToUserKeyTestsContract.sol";
export type { offboardToUserKeyTestsContractSol };
export type { ArithmeticTestsContract } from "./ArithmeticTestsContract";
export type { BitwiseTestsContract } from "./BitwiseTestsContract";
export type { Comparison2TestsContract } from "./Comparison2TestsContract";
export type { MinMaxTestsContract } from "./MinMaxTestsContract";
export type { Miscellaneous1TestsContract } from "./Miscellaneous1TestsContract";
export type { Miscellaneous2TestsContract } from "./Miscellaneous2TestsContract";
export type { ShiftTestsContract } from "./ShiftTestsContract";
export type { StringTestsContract } from "./StringTestsContract";
export type { TransferScalarTestsContract } from "./TransferScalarTestsContract";
export type { TransferTestsContract } from "./TransferTestsContract";
