// @coti-io/coti-contracts/typechain-types/contracts/mocks/utils/mpc/TransferScalarTestsContract.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../../../common";

export declare namespace TransferScalarTestsContract {
  export type AllGTCastingValuesStruct = {
    a8_s: BigNumberish;
    b8_s: BigNumberish;
    a16_s: BigNumberish;
    b16_s: BigNumberish;
    a32_s: BigNumberish;
    b32_s: BigNumberish;
    a64_s: BigNumberish;
    b64_s: BigNumberish;
  };

  export type AllGTCastingValuesStructOutput = [
    a8_s: bigint,
    b8_s: bigint,
    a16_s: bigint,
    b16_s: bigint,
    a32_s: bigint,
    b32_s: bigint,
    a64_s: bigint,
    b64_s: bigint
  ] & {
    a8_s: bigint;
    b8_s: bigint;
    a16_s: bigint;
    b16_s: bigint;
    a32_s: bigint;
    b32_s: bigint;
    a64_s: bigint;
    b64_s: bigint;
  };
}

export interface TransferScalarTestsContractInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "computeAndChekTransfer16"
      | "computeAndChekTransfer32"
      | "computeAndChekTransfer64"
      | "getResults"
      | "transferScalarTest"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "computeAndChekTransfer16",
    values: [
      TransferScalarTestsContract.AllGTCastingValuesStruct,
      BigNumberish,
      BigNumberish,
      boolean,
      BigNumberish
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "computeAndChekTransfer32",
    values: [
      TransferScalarTestsContract.AllGTCastingValuesStruct,
      BigNumberish,
      BigNumberish,
      boolean,
      BigNumberish
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "computeAndChekTransfer64",
    values: [
      TransferScalarTestsContract.AllGTCastingValuesStruct,
      BigNumberish,
      BigNumberish,
      boolean,
      BigNumberish
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "getResults",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "transferScalarTest",
    values: [BigNumberish, BigNumberish, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "computeAndChekTransfer16",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "computeAndChekTransfer32",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "computeAndChekTransfer64",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "getResults", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "transferScalarTest",
    data: BytesLike
  ): Result;
}

export interface TransferScalarTestsContract extends BaseContract {
  connect(runner?: ContractRunner | null): TransferScalarTestsContract;
  waitForDeployment(): Promise<this>;

  interface: TransferScalarTestsContractInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  computeAndChekTransfer16: TypedContractMethod<
    [
      allGTCastingValues: TransferScalarTestsContract.AllGTCastingValuesStruct,
      new_a: BigNumberish,
      new_b: BigNumberish,
      res: boolean,
      amount: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  computeAndChekTransfer32: TypedContractMethod<
    [
      allGTCastingValues: TransferScalarTestsContract.AllGTCastingValuesStruct,
      new_a: BigNumberish,
      new_b: BigNumberish,
      res: boolean,
      amount: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  computeAndChekTransfer64: TypedContractMethod<
    [
      allGTCastingValues: TransferScalarTestsContract.AllGTCastingValuesStruct,
      new_a: BigNumberish,
      new_b: BigNumberish,
      res: boolean,
      amount: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  getResults: TypedContractMethod<[], [[bigint, bigint, boolean]], "view">;

  transferScalarTest: TypedContractMethod<
    [a: BigNumberish, b: BigNumberish, amount: BigNumberish],
    [[bigint, bigint, boolean]],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "computeAndChekTransfer16"
  ): TypedContractMethod<
    [
      allGTCastingValues: TransferScalarTestsContract.AllGTCastingValuesStruct,
      new_a: BigNumberish,
      new_b: BigNumberish,
      res: boolean,
      amount: BigNumberish
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "computeAndChekTransfer32"
  ): TypedContractMethod<
    [
      allGTCastingValues: TransferScalarTestsContract.AllGTCastingValuesStruct,
      new_a: BigNumberish,
      new_b: BigNumberish,
      res: boolean,
      amount: BigNumberish
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "computeAndChekTransfer64"
  ): TypedContractMethod<
    [
      allGTCastingValues: TransferScalarTestsContract.AllGTCastingValuesStruct,
      new_a: BigNumberish,
      new_b: BigNumberish,
      res: boolean,
      amount: BigNumberish
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "getResults"
  ): TypedContractMethod<[], [[bigint, bigint, boolean]], "view">;
  getFunction(
    nameOrSignature: "transferScalarTest"
  ): TypedContractMethod<
    [a: BigNumberish, b: BigNumberish, amount: BigNumberish],
    [[bigint, bigint, boolean]],
    "nonpayable"
  >;

  filters: {};
}
